// vite-plugin-compression/dist/index.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }var __defProp = Object.defineProperty;
var __defProps = Object.defineProperties;
var __getOwnPropDescs = Object.getOwnPropertyDescriptors;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __spreadProps = (a, b) => __defProps(a, __getOwnPropDescs(b));

// src/index.ts
var _path = require('path'); var _path2 = _interopRequireDefault(_path);
var _vite = require('vite');

// src/utils.ts
var _fs = require('fs'); var _fs2 = _interopRequireDefault(_fs);

var isFunction = (arg) => typeof arg === "function";
var isRegExp = (arg) => Object.prototype.toString.call(arg) === "[object RegExp]";
function readAllFile(root, reg) {
  let resultArr = [];
  try {
    if (_fs2.default.existsSync(root)) {
      const stat = _fs2.default.lstatSync(root);
      if (stat.isDirectory()) {
        const files = _fs2.default.readdirSync(root);
        files.forEach(function(file) {
          const t = readAllFile(_path2.default.join(root, "/", file), reg);
          resultArr = resultArr.concat(t);
        });
      } else {
        if (reg !== void 0) {
          if (isFunction(reg.test) && reg.test(root)) {
            resultArr.push(root);
          }
        } else {
          resultArr.push(root);
        }
      }
    }
  } catch (error) {
  }
  return resultArr;
}

// src/index.ts
var _fsextra = require('fs-extra'); var _fsextra2 = _interopRequireDefault(_fsextra);
var _chalk = require('chalk'); var _chalk2 = _interopRequireDefault(_chalk);
var _zlib = require('zlib'); var _zlib2 = _interopRequireDefault(_zlib);
var _debug = require('debug');
var debug = _debug.debug.call(void 0, "vite-plugin-compression");
var extRE = /\.(js|mjs|json|css|html)$/i;
var mtimeCache = new Map();
var exportFn = (options = {}) => {
  let outputPath;
  let config;
  const emptyPlugin = {
    name: "vite:compression"
  };
  const {
    disable = false,
    filter = extRE,
    verbose = true,
    algorithm = "gzip",
    ext = ".gz",
    threshold = 1025,
    compressionOptions = {},
    deleteOriginFile = false,
    success = () => {
    }
  } = options;
  if (disable) {
    return emptyPlugin;
  }
  debug("plugin options:", options);
  return __spreadProps(__spreadValues({}, emptyPlugin), {
    apply: "build",
    enforce: "post",
    configResolved(resolvedConfig) {
      config = resolvedConfig;
      outputPath = _path2.default.isAbsolute(config.build.outDir) ? config.build.outDir : _path2.default.join(config.root, config.build.outDir);
      debug("resolvedConfig:", resolvedConfig);
    },
    async closeBundle() {
      let files = readAllFile(outputPath) || [];
      debug("files:", files);
      if (!files.length)
        return;
      files = filterFiles(files, filter);
      const compressOptions = getCompressionOptions(algorithm, compressionOptions);
      const compressMap = new Map();
      const handles = files.map(async (filePath) => {
        let { mtimeMs, size: oldSize } = await _fsextra2.default.stat(filePath);
        if (mtimeMs <= (mtimeCache.get(filePath) || 0) || oldSize < threshold)
          return;
        let content = await _fsextra2.default.readFile(filePath);
        if (deleteOriginFile) {
          _fsextra2.default.remove(filePath);
        }
        try {
          content = await compress(content, algorithm, compressOptions);
        } catch (error) {
          config.logger.error("compress error:" + filePath);
        }
        const size = content.byteLength;
        const cname = getOutputFileName(filePath, ext);
        compressMap.set(filePath, {
          size: size / 1024,
          oldSize: oldSize / 1024,
          cname
        });
        await _fsextra2.default.writeFile(cname, content);
        mtimeCache.set(filePath, Date.now());
      });
      Promise.all(handles).then(() => {
        if (verbose) {
          handleOutputLogger(config, compressMap, algorithm);
          success();
        }
      });
    }
  });
};
function filterFiles(files, filter) {
  if (filter) {
    const isRe = isRegExp(filter);
    const isFn = isFunction(filter);
    files = files.filter((file) => {
      if (isRe) {
        return filter.test(file);
      }
      if (isFn) {
        return filter(file);
      }
      return true;
    });
  }
  return files;
}
function getCompressionOptions(algorithm = "", compressionOptions = {}) {
  const defaultOptions = {
    gzip: {
      level: _zlib2.default.constants.Z_BEST_COMPRESSION
    },
    deflate: {
      level: _zlib2.default.constants.Z_BEST_COMPRESSION
    },
    deflateRaw: {
      level: _zlib2.default.constants.Z_BEST_COMPRESSION
    },
    brotliCompress: {
      params: {
        [_zlib2.default.constants.BROTLI_PARAM_QUALITY]: _zlib2.default.constants.BROTLI_MAX_QUALITY,
        [_zlib2.default.constants.BROTLI_PARAM_MODE]: _zlib2.default.constants.BROTLI_MODE_TEXT
      }
    }
  };
  return __spreadValues(__spreadValues({}, defaultOptions[algorithm]), compressionOptions);
}
function compress(content, algorithm, options = {}) {
  return new Promise((resolve, reject) => {
    _zlib2.default[algorithm](content, options, (err, result) => err ? reject(err) : resolve(result));
  });
}
function getOutputFileName(filepath, ext) {
  const compressExt = ext.startsWith(".") ? ext : `.${ext}`;
  return `${filepath}${compressExt}`;
}
function handleOutputLogger(config, compressMap, algorithm) {
  config.logger.info(`
${_chalk2.default.cyan("\u2728 [vite-plugin-compression]:algorithm=" + algorithm)} - compressed file successfully: `);
  const keyLengths = Array.from(compressMap.keys(), (name) => name.length);
  const maxKeyLength = Math.max(...keyLengths);
  compressMap.forEach((value, name) => {
    let { size, oldSize, cname } = value;
    const rName = _vite.normalizePath.call(void 0, cname).replace(_vite.normalizePath.call(void 0, `${config.root}/${config.build.outDir}/`), "");
    const sizeStr = `${oldSize.toFixed(2)}kb / ${algorithm}: ${size.toFixed(2)}kb`;
    config.logger.info(_chalk2.default.dim(config.build.outDir + "/") + _chalk2.default.blueBright(rName) + " ".repeat(2 + maxKeyLength - name.length) + " " + _chalk2.default.dim(sizeStr));
  });
  config.logger.info("\n");
}
module.exports = exportFn;
var src_default = exportFn;


exports.default = src_default;
